src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -91,12 +83,12 @@ public void SetTokensLegend(RazorSemanticTokensLegend legend)
 
         cancellationToken.ThrowIfCancellationRequested();
 
-        var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range, _razorSemanticTokensLegend.AssumeNotNull(), colorBackground);
+        var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range, _razorSemanticTokensLegendService, colorBackground);
```

Comment:
I haven't finished reading, so this may be answered later: I expected this to still take the legend instead of the service. Is there a reason for that change?

Reply:
Yeah, its a bit odd. The way I think about is that service is a thing that produces a legend, but it also provides access to the data that contributed to the legend, because the legend is an LSP format that is not very useful. Which does sort of mean the service is also a legend of sorts, just with higher fidelity.